job/deployer.go

```diff
@@ -0,0 +1,150 @@
+package job
+
+import (
+	"context"
+
+	"github.com/google/uuid"
+	"github.com/hashicorp/go-multierror"
+
+	"github.com/odpf/optimus/core/progress"
+	"github.com/odpf/optimus/models"
+)
+
+type deployer struct {
+	dependencyResolver        DependencyResolver
+	priorityResolver          PriorityResolver
+	projectJobSpecRepoFactory ProjectJobSpecRepoFactory
+
+	// scheduler for managing batch scheduled jobs
+	batchScheduler models.SchedulerUnit
+}
+
+func NewDeployer(dependencyResolver DependencyResolver, priorityResolver PriorityResolver, projectJobSpecRepoFactory ProjectJobSpecRepoFactory, batchScheduler models.SchedulerUnit) *deployer {
+	return &deployer{dependencyResolver: dependencyResolver, priorityResolver: priorityResolver, projectJobSpecRepoFactory: projectJobSpecRepoFactory, batchScheduler: batchScheduler}
+}
+
+func (d *deployer) Deploy(ctx context.Context, projectSpec models.ProjectSpec, progressObserver progress.Observer) (deployError error) {
+	// Get all jobs
+	projectJobSpecRepo := d.projectJobSpecRepoFactory.New(projectSpec)
+	jobSpecs, err := projectJobSpecRepo.GetAll(ctx)
+	if err != nil {
+		return err
+	}
+	d.notifyProgress(progressObserver, &models.ProgressJobSpecFetch{})
+
+	// fetch job dependencies
+	jobDependencies, err := d.dependencyResolver.FetchJobDependencies(ctx, projectSpec.ID)
+	if err != nil {
+		return err
+	}
+	d.notifyProgress(progressObserver, &models.ProgressJobSpecDependencyFetch{})
+
+	// enrich with job dependencies
+	jobSpecs, err = d.enrichJobSpecWithJobDependencies(ctx, jobSpecs, jobDependencies)
+	if err != nil {
+		return err
+	}
+	d.notifyProgress(progressObserver, &models.ProgressJobSpecJobDependencyEnrich{})
+
+	// Get all job specs and enrich with hook dependencies
+	jobSpecs = d.enrichJobSpecWithHookDependencies(jobSpecs)
+	d.notifyProgress(progressObserver, &models.ProgressJobSpecHookDependencyEnrich{})
+
+	// Resolve priority
+	jobSpecs, err = d.priorityResolver.Resolve(ctx, jobSpecs, progressObserver)
+	if err != nil {
+		return err
+	}
+
+	// Compile & Deploy
+	jobSpecGroup := models.JobSpecs(jobSpecs).GroupJobsPerNamespace()
+	for _, jobs := range jobSpecGroup {
+		if len(jobs) == 0 {
+			continue
+		}
+		if err := d.batchScheduler.DeployJobs(ctx, jobs[0].NamespaceSpec, jobs, progressObserver); err != nil {
+			deployError = multierror.Append(deployError, err)
+		}
+	}
+
+	return deployError
+}
+
+func (d *deployer) enrichJobSpecWithJobDependencies(ctx context.Context, jobSpecs []models.JobSpec, jobDependencies []models.JobIDDependenciesPair) ([]models.JobSpec, error) {
```

Comment:
we can refactor this method
1. createJobSpecMap can be dissolved into fetchJobSpecsForDependentJobs which can go into the main function call and the map creation will just  create JobSpecMap given jobSpecs and dependentJobSpecs, here we can optimize by grouping the jobs by dependent project and fetching only once. Or why not fetch the jobspecs through a foreign key contraint, then we can avoid this fetch altogether.
2. We will have a function to enrich a single obSpec With JobDependencies, which will be reused here. When ever there is nesting better to break down into multiple functions

Reply:
updated as discussed: not fetching the jobspecs through foreign key constraint (to avoid performance issue), jobSpecMap and enrichment are now being done when fetching job specs with job dependencies.